### ISSUE TITLE
const blocks don't have a recursion limit, causing the compiler to freeze in an infinite loop

### ISSUE THREAD

**Issue by theemathas** [rust-lang/rust#125718]
```rust
const fn foo<T>() {
    const { foo::<&T>() }
}

const X: () = foo::<i32>();
```

The above code causes the compiler to hang, apparently in an infinite loop. I expected it to report an error due to a recursion limit or a timeout.

Note that removing the const block causes the compiler to correctly report an error.
<details>
<summary>Code without the const block</summary>

```rust
const fn foo<T>() {
    foo::<&T>()
}

const X: () = foo::<i32>();
```
This code results in the following error:
```
   Compiling playground v0.0.1 (/playground)
warning: function cannot return without recursing
 --> src/lib.rs:1:1
  |
1 | const fn foo<T>() {
  | ^^^^^^^^^^^^^^^^^ cannot return without recursing
2 |     foo::<&T>()
  |     ----------- recursive call site
  |
  = help: a `loop` may express intention better if this is on purpose
  = note: `#[warn(unconditional_recursion)]` on by default

error[E0080]: evaluation of constant value failed
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^ reached the configured maximum number of stack frames
  |
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<&i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `foo::<i32>`
 --> src/lib.rs:2:5
  |
2 |     foo::<&T>()
  |     ^^^^^^^^^^^
note: inside `X`
 --> src/lib.rs:5:15
  |
5 | const X: () = foo::<i32>();
  |               ^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0080`.
warning: `playground` (lib) generated 1 warning
error: could not compile `playground` (lib) due to 1 previous error; 1 warning emitted
```
</details>

### Meta

`rustc --version --verbose`:
```
rustc 1.80.0-nightly (da159eb33 2024-05-28)
binary: rustc
commit-hash: da159eb331b27df528185c616b394bb0e1d2a4bd
commit-date: 2024-05-28
host: aarch64-apple-darwin
release: 1.80.0-nightly
LLVM version: 18.1.6```

**Comment by fmease** [comment]
NB: This only affects polymorphically recursive calls (i.e., calls where the generic args change at each step).

**Comment by oli-obk** [comment]
Hmm... The immediate solution I can come up with is to use the recursion limit to limit query depth, but that's going to be hard to tune to a good value without breaking random crates (most likely binary crates that we don't see on crater).

**Comment by fmease** [comment]
> The immediate solution I can come up with is to use the recursion limit to limit query depth

That would also solve #114192.

**Comment by saethlin** [comment]
I think there might be something else going on here. If I try to compile this reproducer I get a freeze with growing memory usage, which is what I would expect.

But if I try to ctrl-c the process, I... get a segfault?
```
╭ ➜ ben@archlinux:/tmp
╰ ➤ rustc +nightly ice.rs --crate-type=lib
^Cerror: rustc interrupted by SIGSEGV, printing backtrace

/home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-4f8cb1b2fa29bc89.so(+0x3150996)[0x7de413b50996]
/usr/lib/libc.so.6(+0x3cae0)[0x7de410850ae0]
/home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/libstd-45288dcc88911a1f.so(_ZN4core3fmt5write17hd4702675b19bc4d7E+0x8)[0x7de417308518]
/home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-4f8cb1b2fa29bc89.so(+0x575c169)[0x7de41615c169]

### cycle encountered after 4 frames with period 6
/home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-4f8cb1b2fa29bc89.so(+0x575b40d)[0x7de41615b40d]
/home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-4f8cb1b2fa29bc89.so(+0x575c19b)[0x7de41615c19b]
/home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-4f8cb1b2fa29bc89.so(+0x575b40d)[0x7de41615b40d]
/home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-4f8cb1b2fa29bc89.so(+0x575c19b)[0x7de41615c19b]
/home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-4f8cb1b2fa29bc89.so(+0x575b40d)[0x7de41615b40d]
/home/ben/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-4f8cb1b2fa29bc89.so(+0x575c19b)[0x7de41615c19b]
### recursed 42 times


note: rustc unexpectedly overflowed its stack! this is a bug
note: maximum backtrace depth reached, frames may have been lost
note: we would appreciate a report at https://github.com/rust-lang/rust
help: you can increase rustc's stack size by setting RUST_MIN_STACK=16777216
Segmentation fault (core dumped)
```

**Comment by oli-obk** [comment]
It probably tries to run the Ctrl-C handler and emit a mir interpreter backtrace, but that logic doesn't use `ensure_sufficient_stack` and hits the stack overflow protection. Queries on the other band just allocate more stack when necessary

**Comment by theemathas** [comment]
The original code now segfaults on [playground's](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f4efab1535a00e48865146e07f521862) nightly (2024-07-12 c6727fc9b5c64cefa726). However, I am unable to reproduce the segfault on my computer.

@rustbot label +I-crash +regression-untriaged

```
   Compiling playground v0.0.1 (/playground)
error: rustc interrupted by SIGSEGV, printing backtrace

/playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-7b97f2406392707f.so(+0x342a873)[0x7fb5d823d873]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x14420)[0x7fb5d4cc1420]
/playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-7b97f2406392707f.so(_RNvXNvNtNtCsf1a5ts8v4Q3_12rustc_middle2ty8instance11type_lengthNtB2_7VisitorINtNtCs4rQBiCPslym_13rustc_type_ir5visit11TypeVisitorNtNtB6_7context6TyCtxtE8visit_ty+0x9)[0x7fb5d9816ae3]

### cycle encountered after 3 frames with period 4
/playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-7b97f2406392707f.so(_RNvXNvNtNtCsf1a5ts8v4Q3_12rustc_middle2ty8instance11type_lengthNtB2_7VisitorINtNtCs4rQBiCPslym_13rustc_type_ir5visit11TypeVisitorNtNtB6_7context6TyCtxtE8visit_ty+0xb1)[0x7fb5d9816b8b]
/playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-7b97f2406392707f.so(_RNvXNvNtNtCsf1a5ts8v4Q3_12rustc_middle2ty8instance11type_lengthNtB2_7VisitorINtNtCs4rQBiCPslym_13rustc_type_ir5visit11TypeVisitorNtNtB6_7context6TyCtxtE8visit_ty+0xb1)[0x7fb5d9816b8b]
/playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-7b97f2406392707f.so(_RNvXNvNtNtCsf1a5ts8v4Q3_12rustc_middle2ty8instance11type_lengthNtB2_7VisitorINtNtCs4rQBiCPslym_13rustc_type_ir5visit11TypeVisitorNtNtB6_7context6TyCtxtE8visit_ty+0xb1)[0x7fb5d9816b8b]
/playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-7b97f2406392707f.so(_RNvXNvNtNtCsf1a5ts8v4Q3_12rustc_middle2ty8instance11type_lengthNtB2_7VisitorINtNtCs4rQBiCPslym_13rustc_type_ir5visit11TypeVisitorNtNtB6_7context6TyCtxtE8visit_ty+0xb1)[0x7fb5d9816b8b]
### recursed 63 times

/playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/librustc_driver-7b97f2406392707f.so(_RNvXNvNtNtCsf1a5ts8v4Q3_12rustc_middle2ty8instance11type_lengthNtB2_7VisitorINtNtCs4rQBiCPslym_13rustc_type_ir5visit11TypeVisitorNtNtB6_7context6TyCtxtE8visit_ty+0xb1)[0x7fb5d9816b8b]

note: rustc unexpectedly overflowed its stack! this is a bug
note: maximum backtrace depth reached, frames may have been lost
note: we would appreciate a report at https://github.com/rust-lang/rust
help: you can increase rustc's stack size by setting RUST_MIN_STACK=16777216
error: could not compile `playground` (lib)
```

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23125718.20const.20blocks.20don't.20have.20a.20recursion.20limit.20or.20timeou.E2.80.A6/near/455324962)).

The infinite loop started in 1.78. Before, it just emits a error:
```
error[E0658]: inline-const is experimental
 --> main.rs:2:5
  |
2 |     const { foo::<&T>() }
  |     ^^^^^
  |
  = note: see issue #76001 <https://github.com/rust-lang/rust/issues/76001> for more information
```

@rustbot label -I-prioritize +P-medium -regression-untriaged +regression-from-stable-to-stable

**Comment by RalfJung** [comment]
> The infinite loop started in 1.78. Before, it just emits a error:

That is the release that stabilized const blocks. I assume the infinite loop also happened before if you enable the nightly feature. So it's not really a regression, const blocks always behaved like this.
